### DISCUSSION TITLE
Cannot import router type in create-react-app project

### DISCUSSION THREAD

**Discussion by mordv** [trpc/trpc#965] [Q&A] [answered]
Hi! Great library, happy to use it, api finally looks like it always should have.

However, I have had hard time integrating it with create-react-app. They [won't allow](https://github.com/facebook/create-react-app/issues/9127#issuecomment-844336640) importing any uncompiled code from packages in a monorepo. So I can't import router type to create react client.

Don't have this issue anymore since I don't use cra from now on, but I'm surprised it's not mentioned neither in docs nor in issues.

Not sure if this is an issue, and maybe I'm wrong so I leave it as a discussion.

**Comment by KATT** [comment, marked as answer]
When I started tRPC, my app was CRA+Express (before migrating to Next.js) so I'm pretty sure it should work!

Are you sure you are using `import `**`type`**` { AppRouter } from '...'`?

**Reply by mordv** [reply]
Sorry for the delayed response.
I did use `import type`. 
Well, If no one else faces this problem, it was definitely on my side. Anyway, not gonna return to cra, so there's no problem anymore :)

**Reply by talksik** [reply]
I am still somewhat facing this. The approuter type is in a file within my server package and my client spa app imports it. I think it imports fine but vscode doesn't seem to have intellisense for the trpc object. Intellisense keeps spinning despite trying to restart vscode.

**Reply by KATT** [reply]
Might be a tsconfig issue

**Reply by kuubson** [reply]
> **Note** The repository provided is intended to use mainly absolute imports

@KATT I've created a minimal [reproduction](https://github.com/kuubson/react-vite-trpc) of the issue with the typesafety of the `AppRouter` type. 

To see the issue, go to `apps\client\src\trpc\trpc.ts` and delete the re-export statement (`export * from "@example/server/trpc"`). 

This will cause the `AppRouter` to lose its typesafety when hovering over it.

I would greatly appreciate any suggestions on updating the `tsconfig.json` file to prevent re-exporting server-side code, which currently makes the tRPC types available on the client side.

**Comment by kuubson** [comment]
<details>
  <summary>Sort of outdated message, see below for most recent fix</summary>
@talksik, if you are experiencing issues with IntelliSense on the client side when importing `AppRouter`, ensure that you are not using **absolute paths** on the server side for trpc-related items. This can cause problems. 

If you really want to continue using absolute paths, there is a workaround you can try. You can re-export all the folders, that your trpc config consumes, inside the client folder, such as the /utils, /helpers, and /types. To do this, you can add the following code to the appropriate files:

> **Note** It is possible that this approach might result in leakage of backend functions or helpers to the client, however I am not sure - If anyone has knowledge on this topic, please leave a comment.

#### Server

```js
// server/types/index.ts
export type Role = "USER" | 'ADMIN'

// server/trpc/index.ts

import { initTRPC } from '@trpc/server'

// having this import using absolute path
// and not having the `/types` folder re-exported on the client side `/types` folder
// makes the AppRouter lose intellisense on the client side
import type { Role } from 'types'

const t = initTRPC.context().create()

const { router } = t

const appRouter = router({ hello: t.procedure.mutation(() => ({ role: 'USER' as Role })) })

export type AppRouter = typeof appRouter
```

#### Client

> **Note** Based on my experience, re-exporting these folders can also remove Eslint errors that occur when referencing "@server/trpc" on the client side

```js
// client/types/index.ts
export * from '@server/types'

// client/trpc/index.ts

// if having the above types re-export, the `AppRouter` should not have intellisense issues anymore
import type { AppRouter } from '@server/trpc'

// client/package.json
{
   dependencies: {
      '@server': '*'
   }
}
```

> **Warning** Be careful when re-exporting items on the client side, as the following could occur:

```js
// client/utils/index.ts
export * from '@server/utils.ts'

// server/utils/index.ts

// importing `jsonwebtoken` requires node-related packages like `crypto` etc
// and these packages are not supposed to be used on the client side
// so another workaround for this could be importing `jsonwebtoken` dynamically (see the next code section)
import { verify } from 'jsonwebtoken'

export const verifyToken = verify(...)
```

```js
// server/utils/index.ts - dynamic import

const isWeb = typeof document !== 'undefined'

export const verifyToken = () => {
   if (!isWeb) {
       const { verify } = await import('jsonwebtoken')
       verify(...)
   }
}
```

</details>

**Reply by talksik** [reply]
thanks for helping out! what fixed it for me was reinstalling vscode for some reason.

**Comment by egecavusoglu** [comment]
### Setup
I was having a similar issue with a yarn workspaces monorepo with Expo React Native app and Express backend. 

### Issue 
The issue I believe was that my express app was using absolute imports (`baseUrl` + `paths`) and when I imported `type AppRouter` in the React Native project it had difficulty resolving these imports.

### Solution
Instead of importing from the source code of the backend Express app, I compiled the backend project with `tsc` with `tsconfig.json.compilerOptions.declaration: true` and imported the `AppRouter` type from the compiled `.d.ts` file to fix my problem of intellisense.

**Comment by kuubson** [comment]
<details>
   <summary>The main content of this message is somewhat outdated (again :trollface:), so I've collapsed it. Please direct your attention to the most recent answer ⬇️.</summary/>
   Alright so after a long time of fighting this issue, I've got the solution 🚀

When working with absolute paths, it is all about setting custom `paths` in `tsconfig.json` on the client side ~~and not having a folder named `/trpc` on both server and client side (just use for example `/api` on the client side)~~

## Here are some steps on how to approach it:

1. Export your `AppRouter`. ~~This server side folder can be named `/trpc`~~

```js
// apps\server\trpc\api\router.ts
import { router, userProcedure } from "trpc";

export type AppRouter = typeof appRouter;

const appRouter = router({
  user: userProcedure.query(() => ({ role: "USER" })),
});
```

2. Use `createTRPCReact` + import the `AppRouter` using a relative path to the server. ~~This client side folder cannot be named `/trpc`, otherwise the step 3 won't work and the `AppRouter` will be still `any`~~

```js
// apps\client\src\trpc\trpc.ts
import { createTRPCReact } from "@trpc/react-query";

import type { AppRouter } from "../../../server/trpc/api/router";

export const trpc = createTRPCReact<AppRouter>();
```

3. `tsconfig.json` on the client side

```js
// apps\client\tsconfig.json
{
 "compilerOptions": {
   "baseUrl": "src",
   "paths": {
     "@trpc": ["trpc"], // to get more info about this line, see the comment below this whole message 
     "trpc": ["../../server/trpc"] // ⚠️ crucial fragment - you need to map the import paths, so that typescript does not get confused when importing `AppRouter` ⚠️
   }
 }
}
```

4. Enjoy type-safety! No more need of re-exporting server stuff on the client side - it may leak the business logic - this makes my previous above lengthy message sort of outdated

> **Warning** Mistake I also made was installing the server side package on the client side with `turborepo` - don't make this mistake - just use the relative `import type` for the `AppRouter`

```js
// apps\client\package.json
{
  "name": "@example/client",
  "dependencies": {
    // 🚫 not recommended, I would even say forbidden
    "@example/server": "*"
  }
}
```

## For further reference go to the [template](https://github.com/kuubson/react-vite-trpc) I made for `react` + `vite` + `express` + `tRPC` + `turborepo` where I've managed to make almost everything imported with absolute paths
</details/>

**Reply by kuubson** [reply]
I have dug further and it is possible to have `/trpc` folder on both server and client side.

It comes to setting an additional mapping for client's `/trpc` folder so the final project structure and `tsconfig.json` looks like:

- apps
   - server
      - trpc
   - client
      - trpc
      
 ```js
 {
  "compilerOptions": {
    "baseUrl": "src",
    "paths": {
      "@trpc": ["trpc"], // use @trpc on the client side
      "trpc": ["../../server/trpc"] // it just fulfills the typescript reference for the server stuff
    }
  }
}

 ```

**Reply by clonecorp** [reply]
> Alright so after a long time of fighting this issue, I've got the solution 🚀
> 
> When working with absolute paths, it is all about setting custom `paths` in `tsconfig.json` on the client side ~and not having a folder named `/trpc` on both server and client side (just use for example `/api` on the client side)~
> ...

Wow, thank you for that. That was exactly the answer I was looking for. Could you please explain why doing this is not recommended?
> ```"dependencies": {
>     // 🚫 not recommended, I would even say forbidden
>     "@example/server": "*"
>   }```

I use those to avoid ```../../../../``` this type of thing. But I noticed in a monorepo the aliases are kind of confusing TypeScript and it'd be nice to have a clear way of doing it (aliases within an app as well as devDependencies referring to other apps or packages). Let me know if you have figured it out.

**Reply by kuubson** [reply]
I mean installing other packages the "`"packageName": "*"`" way is totally valid, but the only one exception is declaring the server app inside the client's package.json, because then you sort of expose your business/server logic to the client side

**Reply by kuubson** [reply]
That's why instead of declaring `@example/server: "*"` inside the client's `package.json`, you just import the `AppRouter` using relative paths + in order to make the Intellisense/autocomplete work properly, you need to ensure proper TS references by remapping absolute paths on the client side

**Comment by kuubson** [comment]
## I've just updated the [`react-vite-trpc`](https://github.com/kuubson/react-vite-trpc) monorepo template, which means that most of my previous answers ⬆️ are now somewhat outdated.

Take a look at the [template](https://github.com/kuubson/react-vite-trpc) for the latest solution that utilizes TS references and eliminates the need for paths remapping. 

In reference to the main issue topic, I've also prepared a template for React + CRA: [react-cra-vite](https://github.com/kuubson/react-cra-trpc)
